skbio/stats/power.py

```diff
@@ -870,49 +866,276 @@ def _compare_distributions(test, samples, counts=5, mode="ind", num_iter=1000):
+    p_return = test(samples)
+    if isinstance(p_return, float):
+        num_p = 1
+    elif isinstance(p_return, np.ndarray) and len(p_return.shape) == 1:
+        num_p = p_return.shape[0]
+    else:
+        raise TypeError('test must return a float or one-dimensional array.')
+
+    # Calculates the same counts
+    sample_counts = np.arange(min_counts,
+                              min(max_counts, largest),
+                              counts_interval)
+
+    return ratio, num_p, sample_counts
+
+
+def _identify_sample_groups(meta, cat, control_cats, order, strict_match):
+    """Aggregates samples matches for `control_cats` that vary by `cat`
+
+    Parameters
+    ----------
+    meta : pandas.DataFrame
+        The metadata associated with the samples.
+    cat : str, list
+        The metadata category (or a list of categories) for comparison.
+    control_cats : list
+        The metadata categories to be used as controls. For example, if you
+        wanted to vary age (`cat` = "AGE"), you might want to control for
+        gender and health status (i.e. `control_cats` = ["SEX", "HEALTHY"])
+    order : list
+        The order of groups in the category. This can be used
+        to limit the groups selected. For example, if there's a category with
+        groups 'A', 'B' and 'C', and you only want to look at A vs B, `order`
+        would be set to ['A', 'B'].
+    ctrl_pos : int
+        The location of the smallest group in `order`.
+    strict_match: bool, optional
+        This determines how data is grouped using `control_cats`. If a sample
+        within `meta` has an undefined value (`NaN`) for any of the columns in
+        `control_cats`, the sample will not be considered as having a match and
+        will be ignored when `strict_match` is True. If `strict_match` is
+        False, missing values (NaN) in the `control_cats` can be considered
+        matches.
+
+    Returns
+    -------
+    meta_pairs : dict
+        Describes the categories matched for metadata. The
+        `control_cat`-grouped samples are numbered, correspoinding to the
+        second list in `index`. The group is keyed to the list of sample arrays
+        with the same length of `order`.
+    index : list
+        A list of numpy arrays describing the positions of samples to be drawn.
+        The first array is an index array. The second gives an integer
+        corresponding to the `control_cat`-group, and the third lists the
+        position of the reference group sample in the list of samples.
+
+    """
+
+    # Sets up variables to be filled
+    meta_pairs = {}
+    index1 = []
+    index2 = []
+    i1 = 0
+    i2 = 0
+
+    # Groups the data by the control groups
+    ctrl_groups = meta.groupby(control_cats).groups
+    # Identifies the samples that satisfy the control pairs
+    for (g, ids) in viewitems(ctrl_groups):
+        if not _check_nans(g, switch=True) and strict_match:
+            continue
+        try:
```

Comment:
That requires a list lookup and a separate list traversal.

Just to make sure I understand what's going on:

1) in this forloop, groups of samples are being pulled out
2) it is possible (rarely? frequently?) that some IDs do not exist in some groups
3) an efficient and explicit check is desired to determine if these IDs exist in the groups

If I read this correctly, the current place where a `KeyError` is thrown is either line 1074 or 1075. Is that correct?